File: src/server/routes/audit.ts
// SPDX-License-Identifier: LicenseRef-FSL-1.1-Apache-2.0
import type { FastifyInstance } from "fastify";
import type {
  SessionLifecycleFilters,
  SessionLifecycleRepository,
  SigningRequestFilters,
  SigningRequestsRepository,
} from "../../audit/index.js";

export interface AuditRoutesParams {
  app: FastifyInstance;
  sessionLifecycleRepo?: SessionLifecycleRepository;
  signingRequestsRepo?: SigningRequestsRepository;
}

const SOURCE_VALUES = ["endpoint-auth", "agent-forwarding", "agent-proxy"] as const;
const OUTCOME_VALUES = ["approved", "denied", "expired", "cancelled"] as const;

export function registerAuditRoutes(params: AuditRoutesParams) {
  const { app, sessionLifecycleRepo, signingRequestsRepo } = params;

  if (sessionLifecycleRepo) {
    app.get<{
      Querystring: {
        endpointId?: string;
        from?: string;
        to?: string;
        cursor?: string;
        limit?: string;
      };
    }>("/api/audit/sessions", async (request) => {
      const filters: SessionLifecycleFilters = {
        endpointId: request.query.endpointId,
        from: request.query.from,
        to: request.query.to,
      };
      const limit = parseLimit(request.query.limit);
      const page = sessionLifecycleRepo.list(request.accountId, filters, {
        cursor: request.query.cursor,
        limit,
      });
      return page;
    });
  }

  if (signingRequestsRepo) {
    app.get<{
      Querystring: {
        source?: string;
        outcome?: string;
        from?: string;
        to?: string;
        cursor?: string;
        limit?: string;
      };
    }>("/api/audit/signings", async (request, reply) => {
      const source = parseEnum(request.query.source, SOURCE_VALUES);
      if (source === "invalid") return reply.code(400).send({ error: "invalid source filter" });
      const outcome = parseEnum(request.query.outcome, OUTCOME_VALUES);
      if (outcome === "invalid") return reply.code(400).send({ error: "invalid outcome filter" });

      const filters: SigningRequestFilters = {
        source,
        outcome,
        from: request.query.from,
        to: request.query.to,
      };
      const limit = parseLimit(request.query.limit);
      const page = signingRequestsRepo.list(request.accountId, filters, {
        cursor: request.query.cursor,
        limit,
      });
      return page;
    });

    app.get<{ Params: { id: string } }>("/api/audit/signings/:id", async (request, reply) => {
      const row = signingRequestsRepo.getById(request.accountId, request.params.id);
      if (!row) return reply.code(404).send({ error: "not found" });
      return row;
    });
  }
}

function parseLimit(raw: string | undefined): number | undefined {
  if (!raw) return undefined;
  const n = Number(raw);
  if (!Number.isFinite(n)) return undefined;
  return n;
}

/**
 * Parse and validate a single-value query-string enum filter. Returns:
 *   - `undefined` when the filter wasn't provided
 *   - the value when it's in the allowed set
 *   - the literal `"invalid"` when it isn't (caller emits 400)
 */
function parseEnum<T extends string>(
  raw: string | undefined,
  allowed: readonly T[],
): T | undefined | "invalid" {
  if (raw === undefined || raw === "") return undefined;
  if (!allowed.includes(raw as T)) return "invalid";
  return raw as T;
}

File: internal/rest/audit.go
// SPDX-License-Identifier: LicenseRef-FSL-1.1-Apache-2.0
// Audit REST (port of src/server/routes/audit.ts): keyset-paginated session +
// signing logs, account-scoped, with source/outcome filter validation. Pinned
// by the audit-* goldens.
package rest

import (
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"github.com/rado0x54/shellwatch/internal/audit"
)

// Audit wires the audit read routes.
type Audit struct {
	Sessions *audit.Sessions
	Signings *audit.Signings
}

var (
	sourceValues  = map[string]bool{"endpoint-auth": true, "agent-forwarding": true, "agent-proxy": true}
	outcomeValues = map[string]bool{"approved": true, "denied": true, "expired": true, "cancelled": true}
)

func (a *Audit) Mount(r chi.Router) {
	r.Get("/api/audit/sessions", a.listSessions)
	r.Get("/api/audit/signings", a.listSignings)
	r.Get("/api/audit/signings/{id}", a.getSigning)
}

func (a *Audit) listSessions(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	page, err := a.Sessions.List(r.Context(), accountID(r), audit.SessionFilters{
		EndpointID: q.Get("endpointId"), From: q.Get("from"), To: q.Get("to"),
	}, q.Get("cursor"), parseLimit(q.Get("limit")))
	if err != nil {
		writeErr(w, 500, "internal error")
		return
	}
	writeJSON(w, 200, page)
}

func (a *Audit) listSignings(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	if s := q.Get("source"); s != "" && !sourceValues[s] {
		writeErr(w, 400, "invalid source filter")
		return
	}
	if o := q.Get("outcome"); o != "" && !outcomeValues[o] {
		writeErr(w, 400, "invalid outcome filter")
		return
	}
	page, err := a.Signings.List(r.Context(), accountID(r), audit.SigningFilters{
		Source: q.Get("source"), Outcome: q.Get("outcome"), From: q.Get("from"), To: q.Get("to"),
	}, q.Get("cursor"), parseLimit(q.Get("limit")))
	if err != nil {
		writeErr(w, 500, "internal error")
		return
	}
	writeJSON(w, 200, page)
}

func (a *Audit) getSigning(w http.ResponseWriter, r *http.Request) {
	row, err := a.Signings.GetByID(r.Context(), accountID(r), chi.URLParam(r, "id"))
	if err != nil {
		writeErr(w, 500, "internal error")
		return
	}
	if row == nil {
		writeErr(w, 404, "not found")
		return
	}
	writeJSON(w, 200, row)
}

func parseLimit(raw string) int {
	if raw == "" {
		return 0
	}
	n, err := strconv.Atoi(raw)
	if err != nil {
		return 0
	}
	return n
}
